cloudferrylib/copy_engines/scp_copier.py

```diff
@@ -36,73 +37,95 @@ class ScpCopier(base.BaseCopier):
         """
         Copy a file from source host to destination by scp.
 
-        :param runner: Runner to run a command on source host.
-        :param src_path: Path to a file on source host.
-        :param dst_host: Destination host.
-        :param dst_path: Path to a file on destination host.
+        :param host_src: Source host.
+        :param path_src: Path to a file on source host.
+        :param host_dst: Destination host.
+        :param path_dst: Path to a file on destination host.
         :raise: FileCopyError in case the file was not copied by any reasons.
         """
-        data = {'host_src': runner.host,
-                'path_src': src_path,
-                'host_dst': dst_host,
-                'path_dst': dst_path}
+        cmd = "scp {ssh_opts} {src} {dst}"
+
+        if CONF.migrate.direct_transfer:
+            src = path_src
+            runner = self.runner(host_src, 'src', gateway)
+            run = runner.run
+        else:
+            src = '{src_user}@{src_host}:{src_path}'.format(
+                src_user=CONF.src.ssh_user,
+                src_host=host_src,
+                src_path=path_src,
+            )
+            run = local.run
+        dst = '{dst_user}@{dst_host}:{dst_path}'.format(
+            dst_user=CONF.dst.ssh_user,
+            dst_host=host_dst,
+            dst_path=path_dst,
+        )
+
+        data = {'host_src': host_src,
+                'path_src': path_src,
+                'host_dst': host_dst,
+                'path_dst': path_dst,
+                'gateway': gateway}
 
         retrier = retrying.Retry(
             max_attempts=CONF.migrate.retry,
             predicate=self.verify,
-            predicate_kwargs={'data': data},
-            timeout=0,
+            predicate_kwargs={'data': data}
         )
-        LOG.info("Copying file '%s' to '%s'", src_path, dst_host)
-        cmd = "scp {cipher} -o {opts} {file} {user}@{host}:{path}"
+        LOG.info("Copying file '%s' to '%s'", path_src, host_dst)
         try:
-            retrier.run(runner.run, cmd, opts='StrictHostKeyChecking=no',
-                        file=src_path, user=CONF.dst.ssh_user, host=dst_host,
-                        path=dst_path, cipher=ssh_util.get_cipher_option())
+            retrier.run(run, cmd.format(
+                ssh_opts=ssh_util.default_ssh_options(),
+                src=src,
+                dst=dst),
+                        capture_output=False)
         except retrying.MaxAttemptsReached:
+            self.clean_dst(data)
             raise base.FileCopyError(**data)
 
     def transfer(self, data):
-        src_host = data['host_src']
-        src_path = data['path_src']
-        dst_host = data['host_dst']
-        dst_path = data['path_dst']
+        host_src = data['host_src']
+        path_src = data['path_src']
+        host_dst = data['host_dst']
+        path_dst = data['path_dst']
+        gateway = data.get('gateway')
```

Comment:
These conversions from `data` to variables and back to dict look implicit and unreadable, maybe would be better to pass `src_host`, `src_path`, `dst_host`, and `dst_path` directly to each `transfer` method?